http2: optimize neverIndex lookup with SafeSet

Ayoub-Mabrouk opened the pull request:
Optimizes `neverIndex` lookup by replacing `Array.includes()` with `SafeSet.has()` for O(1) lookup performance.

- SafeSet instead of Array: O(1) hash-based lookup vs O(n) linear search
- `for` loop instead of `.map()`: Direct Set construction avoids intermediate array allocation

ljharb commented:
it's also worth noting that v8 has long optimized `.map` in this case so that there's no intermediate array allocation to speak of.

Ayoub-Mabrouk commented on lib/internal/http2/util.js:
```diff
@@ -761,7 +761,10 @@ function buildNgHeaderString(arrayOrMap,
-  const neverIndex = sensitiveHeaders.map((v) => v.toLowerCase());
+  const neverIndex = new SafeSet();
+  for (let i = 0; i < sensitiveHeaders.length; i++) {
+    neverIndex.add(sensitiveHeaders[i].toLowerCase());
```
The usage of primordials is not advised in some cases like http2
I guess it's been stated here

https://github.com/nodejs/node/blob/main/doc/contributing/primordials.md

ljharb replied:
then wouldn't SafeSet, also a primordial, be discouraged?

ovflowd reviewed (approved):
Looks good to me, won't be a massive performance improvement, but going from O(n) to O(1) is a good win here and the diff is small enough and clean enough.

SGTM.

ovflowd commented:
> it's also worth noting that v8 has long optimized `.map` in this case so that there's no intermediate array allocation to speak of.

Jordan has a point here, but the line of optimization is soo thin that in this scenario _maybe_ relying on explicit optimizations is worth it. That said, these are "technicalities" at best, both the old diff and the proposed changes should work perfectly fine without performance issues or observable performance difference. **(Jordan knows way better than me here tho)**

Ayoub-Mabrouk commented:
That's a great point about V8's .map() optimization, thank you @ovflowd for clarifying! I went with the explicit for loop since I've seen feedback from collaborators about preferring explicit loops over iteration-based methods as mentioned in primordials.md. I'm just trying to follow the contribution guidelines as closely as possible, but I appreciate you both pointing out the nuances here, you and @ljharb  clearly have way more experience with these V8 optimization details than I do. Happy to adjust to whatever approach you think is best!

ljharb commented:
the _iterator protocol_ should be staunchly avoided, but array iteration methods are wonderful :-) a for loop is fine, and more obvious that it's performant, it's just not _more_ performant than .map and friends.

ovflowd commented:
@pimterry quite insightful, had no idea this specific Set creation was that much slower. If that's the case, it completely diminishes the point of this PR as the creation itself would worse the overall performance.

I'm unfamiliar with the defaults here and the callstack chain of this method so... appreciate for clarifying.

Ayoub-Mabrouk commented:
> Sorry, but I don't think this will be an improvement in practice.
> 
> I agree with all the discussion in terms of the general case for this kind of logic, it's a good optimization approach most of the time, but it misses the reality here: in 99.99% of cases this array is empty. Nearly nobody uses explicit sensitive header configuration - the defaults within nghttp2 itself already cover authorization headers & many cookies, and I don't think many users are aware of it anyway. In the few remaining cases it'll usually contain one single element. O(N) vs O(1) isn't relevant.
> 
> In practice, I expect the overhead of set creation vs array creation will dominate this change, and the impact on the vast majority case outweighs the performance boost in the extremely niche scenario where this array is large. It's not a big difference in either case, but I do think overall this'll make http2 fractionally slower for almost all real-world usage.
> 
> Doing a very rough local benchmark, I see empty set creation being about 20x slower than empty array creation, empty set.has being about 10x slower than empty array;includes, and set.has vs array.includes taking about the same amount of time for single-element arrays. There's little benefit until the set gets larger (and this one nearly never will).
> 
> Open to benchmarks or arguments against if I've missed something of course. If we do want to improve this, imo the best gains would be in optimizing this for the empty-array case, to skip the map & includes entirely there (if we can show a benchmark that actually improves things in that very common scenario).

Thanks for looking into this! A few thoughts:
The set creation happens once at module load, not per request - even if it’s slower to create, that’s a one-time startup cost. The .has() lookup runs in the hot path every time buildNgHeaderString is called, which is where the O(1) vs O(n) matters.
Re: the benchmarks - I’d genuinely be interested to see them if you have numbers to share. “Rough local benchmarks” are hard to evaluate without the actual data and methodology. If there’s a real performance regression in the common case, I’m definitely open to seeing it.
Also worth noting that users who do configure sensitive headers (security-focused apps, enterprise setups, etc.) would benefit from this. Optimizing only for the empty array case means ignoring their use case entirely.
If you think this makes things slower, would you be able to share benchmark results showing that? Otherwise I’m not sure how to address the concern without concrete data.

ovflowd commented:
Tend to completely agree with @pimterry here; Even more regarding the Big-O notations for the intended use cases here. The overhead of using a Set here pretty much destroys the benefits brought for the loop for the use cases. You could simply switch the Set for an Array and keep the call with .includes() or even do cheap lazy comparison: `array[0] === yourValue || array.includes(yourValue)`

In the end my general advice with this PR and your previous PR are: Before submitting changes, study the use cases of the code, benchmark your changes and ponder if your changes bring any realistic/observable improvement in terms of performance versus the current implementation. Even if the changes are well intended, you're consuming maintainers time for what seems to be changes that aren't really needed. (Not trying to discourage you from contributing, but encouraging to do more homework before landing a PR) -- That said, I think this specific PR is highly educational for all parties involved.

Ayoub-Mabrouk commented:
Thanks for the feedback @ovflowd , and thanks for the approval! I appreciate you taking the time to review.

I wanted to clarify something - you mentioned agreeing with @pimterry  benchmarks, but I noticed a few issues with those benchmarks that might affect the conclusions:
1. There's a bug in the code - the `Set.has()` benchmark uses wrong variables (`endSet - startSet` instead of `endSetHas - startSetHas`)
2. The benchmarks don't test the optimized approach - they test always creating a Set, but the new commit only creates SafeSet when `sensitiveHeaders.length > 0`
3. They don't simulate the actual per-request pattern with realistic header counts

The original PR had the issue you're concerned about - it always created SafeSet even for empty arrays. But the new commit fixes that: it only creates the SafeSet when `sensitiveHeaders.length > 0`. For empty arrays (the vast majority of requests), it uses `null` and does a simple null check instead of creating any data structure.

> The overhead of using a Set here pretty much destroys the benefits

This is exactly why the new commit avoids Set creation for empty arrays. The optimization directly addresses this concern.

> You could simply switch the Set for an Array and keep the call with .includes() or even do cheap lazy comparison: array[0] === yourValue || array.includes(yourValue)

Your suggestion still requires array creation. The optimized approach avoids any data structure creation for empty arrays, which is the common case. Also, SafeSet is already used in this file (line 764), so this isn't about introducing it - it's about optimizing when it's created.

I've shared benchmark results in my response to @pimterry  that show ~68% improvement for the common case. The benchmark simulates the actual per-request pattern and tests the optimized approach specifically.

If you get a chance, I'd love your thoughts on whether the benchmark approach makes sense, or if there's anything I should adjust.

Regarding your mention of my previous PR - I'd prefer to keep the discussion focused on this specific change. Each PR should be evaluated on its own merits, and I've learned a lot from the feedback on both.

Thanks again for the review!

ovflowd commented on lib/internal/http2/util.js:
```diff
@@ -787,7 +793,7 @@ function buildNgHeaderString(arrayOrMap,
       singles.add(key);
     }
-    const flags = neverIndex.includes(key) ?
+    const flags = (neverIndex !== null && neverIndex.has(key)) ?
```
```suggestion
    const flags = (neverIndex && neverIndex.has(key)) ?
```

Ayoub-Mabrouk commented:
Thanks for the feedback @ovflowd , I appreciate you taking the time.

> I'm not sure how I feel with your large AI-generated(assisted) comments. It feels to me you're stretching a bit.

I understand the concern. English isn't my first language, so I do spend extra time making sure my explanations are clear and thorough - perhaps that comes across differently than intended. I spent about half a day working on a realistic benchmark and writing the analysis. The code and findings are mine. If there's something specific in the technical content that's incorrect, I'd be happy to address it.

> I'm not sure if I'm being clear enough, apologies, just trying to provide feedback that your time would be rather better spent on fixing current real bugs or performance issues within the repo.

This is a performance issue. It's a micro-optimization, but it's in a hot path - `buildNgHeaderString` runs on every HTTP/2 request. As Donald Knuth said, "We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%." This is in that 3%, a function called millions of times in production.

I appreciate the feedback. I'm a bit confused though - you approved the PR, then mentioned it's insignificant. If it's truly insignificant, I'm not sure why there's concern about the approach? And if the benchmark approach is wrong, I'd be happy to adjust it - what would you prefer instead?

> I agree this PR is what'd we call a "good first issue" to get familiar with the codebase, but if the differences are diminished it kinda defeats the purpose of this PR.

I appreciate that perspective. I'm here to learn, and I'm genuinely open to any learning opportunity. I'd happily spend my whole day trying to find the truth - that's part of how I learn. I'm not trying to waste your time, and I apologize if it feels that way. I just find it helpful to understand the "why" behind conclusions, especially when I can test them. If the differences are truly diminished, I'd love to see that demonstrated with actual test results - that would be a great learning opportunity for me. That's how I learn best - by understanding the reasoning, not just accepting statements.

> I'm having issues understanding your benchmark and how it assesses the code, not sure if that benchmark is reliable. It is also using Set which is different from node's internal SafeSet

The benchmark simulates the actual per-request pattern - `buildNgHeaderString` is called once per request, and it processes all headers in that request. It tests empty arrays (99%+ of cases), one header, and five headers, with realistic header counts (10-20 headers per request). I tried to make it reflect what actually happens in real-world usage. SafeSet is a wrapper around Set, so the performance characteristics are identical. The benchmark code is in the PR comments if you'd like to run it and share your results.

> I bet this method is covered by out test suite

The test suite validates correctness, not performance. That's why I created a benchmark.

> In the end of the day, a theoretical benchmark of these changes on userland, won't test the actual real change/impact these changes make on the actual compiled Node.

Fair point. Have you tried running the benchmark? It's in the PR comments. I'd be curious to see your results if you run it, especially since you mentioned:

> Tend to completely agree with @pimterry here

But @pimterry's benchmark has a syntax bug (uses `endSet - startSet` instead of `endSetHas - startSetHas` for the Set.has() test) and doesn't test the optimized approach at all. It tests always creating a Set, but the new commit only creates SafeSet when `sensitiveHeaders.length > 0`.

If you could run the benchmark and share results, that would be really helpful. Right now we have:
- Your assessment that it's insignificant
- @pimterry's benchmark (which has a bug and doesn't test the optimized approach)
- My benchmark (which I'd love for you to try)

I'm open to being wrong, but I'd prefer to base that on actual test results rather than assumptions.

I've put real work into this - the benchmark, the analysis, addressing feedback. If the PR should be closed, that's fine, but I'd prefer we base that decision on actual test results rather than assumptions. I'm happy to adjust the approach if you have specific suggestions - I'm here to learn and improve.

ovflowd commented:
> I understand the concern. English isn't my first language, so I do spend extra time making sure my explanations are clear and thorough - perhaps that comes across differently than intended.

Sure, as I shared, it was the feeling I got from reading the messages. I do of course appreciate you taking your time to go through this.

> I appreciate the feedback. I'm a bit confused though - you approved the PR, then mentioned it's insignificant. If it's truly insignificant, I'm not sure why there's concern about the approach? And if the benchmark approach is wrong, I'd be happy to adjust it - what would you prefer instead?

It is seen as an _insignificant_ change in the "grand order of things", what I mean comes by the fact that it is a small change with questionable gains. I can approve a PR even if I believe on the prior quote I just shared. I'm just not -1, I'm neutral.

> The benchmark simulates the actual per-request pattern - buildNgHeaderString is called once per request, and it processes all headers in that request. It tests empty arrays (99%+ of cases), one header, and five headers, with realistic header counts (10-20 headers per request). I tried to make it reflect what actually happens in real-world usage. SafeSet is a wrapper around Set, so the performance characteristics are identical. The benchmark code is in the PR comments if you'd like to run it and share your results.

> But @pimterry's benchmark has a syntax bug (uses endSet - startSet instead of endSetHas - startSetHas for the Set.has() test) and doesn't test the optimized approach at all. It tests always creating a Set, but the new commit only creates SafeSet when sensitiveHeaders.length > 0.

My commentary is about his take on this PR, not his benchmark.

> The benchmark simulates the actual per-request pattern - buildNgHeaderString is called once per request, and it processes all headers in that request. It tests empty arrays (99%+ of cases), one header, and five headers, with realistic header counts (10-20 headers per request). I tried to make it reflect what actually happens in real-world usage. SafeSet is a wrapper around Set, so the performance characteristics are identical. The benchmark code is in the PR comments if you'd like to run it and share your results.

The problem with such claims, is that you gotta then run benchmarks that are then also realistically and not synthetic. For example, have you checked how an actual run-in-the-mill express.js or pure http2 web-server making requests and handling them benefits from this? Since your benchmark runs in the userland the optimizations that happen there are somewhat different from what happens inside Node. The current code (before your changes) ran in userland will perform differently when executed on userland which is what your benchmark is doing. Could you actually compile Node and then actually benchmark the calls with http2 and see if there's any realistic difference? (you can also attach a debugger)

You're also saying that the performance comes from the amount of headers, as it goes through all headers, but again, the for loop only iterates the sensitive headers which in most scenario are empty or 1 so it won't even iterate on said for loop meaning whatever change you're doing isn't even accounting for anything.

At this point, it is hard to really keep engaging in a faithful discussion about these changes, because your arguments just don't make any sense (to me)... The part that sucks is you're describing your changes as they're that weak link of 3% that will change the performance of http2 drastically when the reality is, these changes on 99% won't even have any effect by @pimterry's claims.

Anyhow, I will disengage from this discussion. I approved the PR, and @pimterry gave a clear -1, he's the one you need to convince, not me. Take care!

Ayoub-Mabrouk commented:
> Apologies, I misquoted you.

No worries at all! 😊

> Yeah, the current commit looks good. But then we fallback to one of my arguments of: Should we merge it? does it bring measurable improvements? Because if it doesn't, "don't fix what ain't broken", avoiding changing code where the gains are minimal... we shouldn't do that. Say, that this change brings visible improvements, absolutely merge it. My whole take is on: This whole time we're spending here could be directed towards something else, but you decide how you spend your time ;)

That's a fair question. The benchmark shows ~68% improvement for the empty array case (which is 99%+ of requests), but you're right that the question is whether that translates to visible improvements in real-world usage. I understand the "don't fix what ain't broken" principle - that's why I created the benchmark to try to quantify the impact. Whether it's worth merging is ultimately up to the maintainers' judgment. I appreciate you taking the time to engage with the technical details.

> Sure thing, I also appreciate for the respectdul conversation. Apologies if I might sound harsh at times, I'm also not a native speaker and finding the right words at times can be challenging.

I completely understand - I'm not a native speaker either, so I know how challenging it can be to find the right words, especially in technical discussions. I appreciate the respectful conversation too, and I didn't find you harsh at all. Thanks for all the feedback! 🙏

pimterry commented:
@Ayoub-Mabrouk Ok, you're clearly committed to this and I think there are some practical improvements here, let's do it :+1:

The empty case optimization is a good move - that is likely to have a much larger effect in practice (a little speed-up for every single HTTP/2 header!) and I can believe the has/includes difference is close enough to be equivalent for the single-element case, and worthwhile for larger sensitive headers arrays if that ever does come up, so it's effectively free upside (though I do still expect that upside case to be _very_ rare).

Still, it would be good to benchmark the full behaviour to confirm this impact applies in practice, rather than benchmarking raw components separately. Can you write up a benchmark script that directly runs the `buildNgHeaderString` method with a few different inputs, and then run it using Node built with & without this change, and share the script & results here? You'll need to run node with `--expose-internals` and then access it with `require('internal/http2/util').buildNgHeaderString`.

If you want to benchmark full the HTTP/2 request flow that would be interesting too, but I suspect that will be much more noisy so it'll be challenging.

If that benchmark shows concrete improvement across the 3 input scenarios then I'm happy to merge this.
